File: python/Stack/739. Daily Temperatures.py
'''
Monotonically Non-increasing Stack:
Time: O(n)
Space: O(n)
'''

class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        res = [0] * len(temperatures)
        stack = []

        for i, t in enumerate(temperatures):
            while stack and t > temperatures[stack[-1]]:
                prev = stack.pop()
                res[prev] = i - prev

            stack.append(i)

        return res

File: javascript/Stack/739. Daily Temperatures.js
/**
 * @param {number[]} temperatures
 * @return {number[]}
 */
var dailyTemperatures = function(temperatures) {
    const res = new Array(temperatures.length).fill(0);
    const stack = [];

    temperatures.forEach((t, curIdx) => {
        while (temperatures[stack?.[stack.length-1]] < t) {
            prevIdx = stack.pop();
            res[prevIdx] = curIdx - prevIdx;
        }
        stack.push(curIdx);
    });
    return res;
};